File: docs/JAMIA/comparisons/RPython/TM/12844529_groupby_drops_object_cols.py
# ID: SO#12844529
# Language: Python
# Bug class: Column ref
# Runtime consequence: DC
# In study: Yes
# Inclusion rationale: groupby aggregate silently drops object-dtype columns. Output missing columns, no warning.
# Reproduction status: Live
# Type system catch: Schema type tracks column presence

import pandas as pd

# Data loaded as object dtype (common when reading from CSV without type inference)
df = pd.DataFrame({
    'city': ['NYC', 'NYC', 'LA', 'LA'],
    'temp': ['72', '75', '80', '82'],  # strings, not numbers
    'humidity': ['45', '50', '60', '55'],  # strings, not numbers
})

print("dtypes:")
print(df.dtypes)
print()

# BUG: groupby().mean() silently drops non-numeric columns — returns empty
result = df.groupby('city').mean(numeric_only=True)
print("After groupby().mean() — all value columns silently dropped:")
print(result)
print(f"Result shape: {result.shape}")
# Empty DataFrame with 0 columns. No error. No warning.

File: docs/JAMIA/comparisons/RPython/TM/12844529_groupby_drops_object_cols.ts
/**
 * RPython SO#12844529 — No numeric types to aggregate - change in groupby() behaviour?
 * Effect: DC (silent data corruption)
 * Bug class: Implicit column selection
 *
 * In pandas, groupby().mean() silently drops object-dtype columns, returning
 * an empty DataFrame if all columns are strings.
 *
 * In tidy-ts, s.mean() on a string[] is a compile-time error. You must convert first.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { city: "NYC", temp: "72", humidity: "45" },
  { city: "NYC", temp: "75", humidity: "50" },
  { city: "LA", temp: "80", humidity: "60" },
  { city: "LA", temp: "82", humidity: "55" },
]);

// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("temp"));

// Fix: explicit conversion
const numeric = df.mutate({
  temp: (r) => parseFloat(r.temp),
  humidity: (r) => parseFloat(r.humidity),
});

const result = numeric.groupBy("city").summarize({
  mean_temp: (g) => s.mean(g.extract("temp")),
  mean_humidity: (g) => s.mean(g.extract("humidity")),
});

result.print();
